### PR TITLE
Update empty_loop documentation/message.

### LINKED ISSUE #6161: Lint: enable empty_loop for no_std crates
Right now, we have a style lint against the following code:
```rust
loop {}
```
Note that this is no longer undefined behavior (since https://github.com/rust-lang/rust/pull/77972).

Currently, there is no way to have a `no_std` binary or library use the `empty_loop` lint.
Even adding `#![deny(clippy::empty_loop)]` to a crate does nothing. This is because #5086 (fixing #3746), made the lint essentially not exist for `no_std` crates.

Furthermore, even in `no_std` crates, using a "hot" deadloop is almost always not what you want (as it burns a bunch of CPU, and can cause crashes). Even on a `no_std` target, a user should either:
  - Panic
  - Call a platform-specific `halt` or `pause` intrinsic.

Note, [`core::sync::atomic::spin_loop_hint`](https://doc.rust-lang.org/core/sync/atomic/fn.spin_loop_hint.html) is not a good recommendation here, as `pause` isn't meant for dead-loops (it's designed for spin-locks). See https://github.com/rust-lang/rust/pull/77924

In my opinion, we should:
  - Reenable the `empty_loop` lint for `no_std` crates.
  - Have a `no_std` specific help message.
  - Disable this lint automatically in [`#[panic_handler]`](https://doc.rust-lang.org/nomicon/panic-handler.html)s
    - As an alternative, we could just disable this lint for `no_std` binaries (while keeping it on for `no_std` libraries).
    - The only uses of deadloops in the [Rustonomicon](https://github.com/rust-lang/nomicon) are inside of `#[panic_handler]`.
  
Our documentation for fixing this (on `std` and `no_std`) is part of #6162  

CC (people involved with this change last time): @therealprof, @oli-obk, @eddyp, @Areredify, @phansch

### PR DESCRIPTION
**Pull request by josephlr**
Originally part of #6161, but now this PR only deals with `std` crates

This change:
  - Updates the `std` message .
  - Updates the docs to mention how the busy loops should be fixed
    - Gives examples of how to do this for `no_std` targets
  - Updates the tests/stderr files to test this change.

changelog: Update `empty_loop` lint documentation

### CODE REVIEW

**Comment by phansch** [comment]
Thanks for the PR! The implementation looks good to me but I'm not using no_std myself, so I've asked on Zulip for a second pair of eyes.

**Comment by Lokathor** [comment]
Using `spin_loop_hint` does *not* resolve the LLVM bug.

Further, `spin_loop_hint` is likely to do nothing at all on any sort of `no_std` where there's actually no threading available (as opposed to a `no_std` lib that is then called by `std` using code that manages the threads).

**Comment by jyn514** [comment]
Yeah really what this needs is a `std::hint::make_forward_progress()` intrinsic, but that doesn't exist AFAIK. Maybe you can emulate it with a volatile read/write?

**Comment by Lokathor** [comment]
You can.
```rust
let x = 0;
unsafe { (&x as *const i32).read_volatile() };
```
All the "forward progress" you'd ever need.

**Comment by jyn514** [comment]
Nice, that looks about right.

@josephlr either way, this should not have a `spin_loop_hint()` suggestion; I would recommend the volatile read unconditionally since this pops up for both std and no-std.

**Comment by jyn514** [comment]
> In non-no_std code, you usually want to sleep or block, so I think we should not suggest the volatile read there.

Sure, if you want to suggest `thread::sleep(...)` instead that sounds better.

>  For no_std that could work, but personally I'm a bit concerned about suggesting unsafe instead of the (supposedly) safe code of loop {}.

🤷 this is the situation we're in. It's better to have visible unsafe than invisible unsoundness. Right now `loop {}` looks the same to LLVM as `unsafe { unreachable_unchecked() }` which is clearly worse.

**Comment by ebroto** [comment]
>It's better to have visible unsafe than invisible unsoundness

I guess you're right. We can always suggest it as a last resort.

**Comment by Lokathor** [comment]
no_std has no portable way to sleep. you can't suggest thread::sleep, because there's no such thing.

It's maybe a drag, but as jyn514 said, this is the world we live in.

**Comment by jonas-schievink** [comment]
You can also suggest compiling with `-Zinsert-sideeffect` on nightly

**Comment by therealprof** [comment]
How about making sure that `spin_loop_hint` always has a side effect? 

The documentation clearly states its intended use:
```
Signals the processor that it is inside a busy-wait spin-loop ("spin lock").

Upon receiving spin-loop signal the processor can optimize its behavior by, for example, saving power or switching hyper-threads.
```

If it cannot be used for an empty loop I'd consider that a bug in `spin_loop_hint`.

BTW: It works just fine for the majority of Rust supported embedded systems (running on  ARM Cortex-M MCUs) by turning into a `yield` instruction.

**Comment by therealprof** [comment]
> It's a hint. A hint never changes the correctness (or not) of a program.

The program is correct with or without the hint. However since empty loops are frowned upon and miscompiled by the compiler there should be a proper and official way to declare an intentionally empty loop.

> For example, I do my embedded work on the GBA, and spin_loop_hint does nothing because the ARM7TDMI CPU of a GBA has no built in assembly instruction for a power saving mode like that.

Fair enough, but what would be the issue using a `nop` instead? The `yield` instruction on most Cortex-M MCUs is also just a fancy way to encode a `nop`.

**Comment by Lokathor** [comment]
The fact that the codegen will miscompile the program means that the program is *not* correct with or without the hint.

The *intended* semantics of `loop{}` is "that's fine" the *actual* semantics of `loop{}` is "LLVM will burn you". We have to live by the semantics we get.

The official, unstable, way to handle this problem is the `-Zinsert-sideeffect` flag that `jonas-schievink` mentioned.

The *stable* way to handle this is to volatile read a stack variable, like I said.

> Fair enough, but what would be the issue using a nop instead?
Well, because that's not a loop.

**Comment by therealprof** [comment]
> The fact that the codegen will miscompile the program means that the program is _not_ correct with or without the hint.
> The _intended_ semantics of `loop{}` is "that's fine" the _actual_ semantics of `loop{}` is "LLVM will burn you". We have to live by the semantics we get.

You're sort of contraticting yourself. A program following the _intended_ semantics is a correct program. Linting about difference actual vs intended semantics and telling the user how to bridge the gap is exactly the job of clippy.

> The official, unstable, way to handle this problem is the `-Zinsert-sideeffect` flag that `jonas-schievink` mentioned.
> The _stable_ way to handle this is to volatile read a stack variable, like I said.

 Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.

> > Fair enough, but what would be the issue using a nop instead?

> Well, because that's not a loop.

This is about preventing the loop from being turned into mush by creating a side effect that forces the compiler to keep it. Of course the `nop` goes into `loop` and does not replace it...

**Comment by Lokathor** [comment]
> Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.

Anything else is UB.

Any volatile isn't "weird unrelated unsafe magic", every embedded developer needs to understand when to use volatile.

**Comment by therealprof** [comment]
> > Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.
 
> Anything else is UB.

It's not supposed/documented to be. And at the moment there's also no lint for it which is what this issue is all about. People are regularly running into problems with miscompilations due to this.

> And volatile isn't "weird unrelated unsafe magic", every embedded developer needs to understand when to use volatile.

The matter of fact is a lot of people don't know it and they shouldn't have to (if we did our job right and abstracted all of that away which we do!). Requiring people to know things like that is a substantial hazard to Rust adoption and requiring such a magic block of unsafe codeit in **every application** is just ridiculous.

**Comment by josephlr** [comment]
Looking at all the comments it seems like we will need to cover multiple use-cases including `no_std` applications and `no_std` libraries. I also agree with the above comments that our advice should make sense _even if we didn't have this LLVM bug_. This seems reasonable, as a hot CPU deadloop is almost always bad style.

Here would be my plan:
  - Remove references to `spin_loop_hint`, as it doesn't (yet) do the right thing for all targets.
    - Once `core` is fixed to make `spin_loop_hint` have a side-effect on all platforms, we recommend that.
  - Update the `no_std` prompt to say something like:
    ```
    empty `loop {}` detected. You may want to either use `panic!()`
    or add a platform-specific yield intrinsic to the loop body.
    ```
    - This handles the `no_std` library case (where libraries should panic instead of deadlooping)
    - This also handles the `no_std` application case (where a binary should do _something_ in it's panic handler other than deadlooping).
  - Update the docs to mention various remediations:
    - Mention that most deadloops outside of the panic handler should panic instead of deadlooping
    - Call out the LLVM bug
    - Calling some platform intrinsic
    - Using `-Zinsert-sideeffect`
    - spin_loop_hint

**Comment by jyn514** [comment]
> a hot CPU deadloop is almost always bad style

This is exactly why I think this should _not_ treat this as if the LLVM bug should not exist. Linting on `loop {}` because it burns up the CPU is performance issue; linting on it because it causes a miscompilation is a correctness issue. If we lump them together I think people will just ignore the lint.

**Comment by josephlr** [comment]
> Linting on `loop {}` because it burns up the CPU is performance issue; linting on it because it causes a miscompilation is a correctness issue.

That makes sense. Are you suggesting changing the lint category from "Style" to "Correctness"?

**Comment by Lokathor** [comment]
@therealprof
> It's not supposed/documented to be.

But it **is** how things are. There's an open bug about it. No amount of "it shouldn't be that way" will change the facts.

Sometimes Rust just *isn't good yet* and you have to wait for it to improve. If you want this fixed, help patch LLVM to allow side-effect free loops, or help move `insert-sideeffect` to be on by default and stable.

Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

@josephlr 
> Once core is fixed to make spin_loop_hint have a side-effect on all platforms, we recommend that.

Is there even an issue about this ever happening? Because I don't think this will otherwise happen.

**Comment by josephlr** [comment]
> Is there even an issue about this ever happening? Because I don't think this will otherwise happen.

I'm working on a patch for `libcore` right now.

**Comment by therealprof** [comment]
> But it **is** how things are. There's an open bug about it. No amount of "it shouldn't be that way" will change the facts.

Not sure where you got this notion from. We're having this discussion on the clippy repository because we want to tell people about the facts and how to remedy them (in a sane way).
 
> or help move `insert-sideeffect` to be on by default and stable.

I am not convinced that this is a solution, at best it's a workaround that has the potential to pessmize generated code even further.

> Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

I fully agree.

**Comment by jyn514** [comment]
> Are you suggesting changing the lint category from "Style" to "Correctness"?

**Comment by flip1995** [comment]
After reading all the comments in this thread and most of the comments in https://github.com/rust-lang/rust/issues/28728 I conclude:

1. The lint should be `correctness`, since `loop {}` can produce UB. We should also state this in the lint error message, because:
    > Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

2. It should also lint for `no_std` crates

3. Suggestion is kind of hard, since there is no good solution, that works on all platforms. So we could do the following:
    a. Mention every ~~solution~~ workaround in the lint message (meh)
    b. Try to detect the platform and suggest the best workaround (how?, which?)
    c. Just point to the documentation and write up a summary of the workarounds there

3.a. seems to be too much (useless) information in the lint message.
3.b. seems hard to do. @josephlr suggestion

> Update the `no_std` prompt to say something like:
> ```
> empty `loop {}` detected. You may want to either use `panic!()`
> or add a platform-specific yield intrinsic to the loop body.
> ```
> 
> * This handles the `no_std` library case (where libraries should panic instead of deadlooping)
> * This also handles the `no_std` application case (where a binary should do _something_ in it's panic handler other than deadlooping).

may work for `no_std`, but I wouldn't know what to do and would have to read some documentation anyway.

So I would go with 3.c. What workarounds should we suggest? (please add to the list, if I forgot something)

- Use `-Zinsert-sideeffect`. From reading https://github.com/rust-lang/rust/issues/28728, this seems to me to be the correct solution for this, but requires unstable features, which may not be possible for some people
- Use some volatile read operation. In https://github.com/rust-lang/rust/issues/28728 it was suggested to add `unsafe {llvm_asm!("" :::: "volatile")}`, which prevents this bug, but again needs the unstable feature `llvm_asm`. The `read_volatile()` produces unnecessary code though (including an allocation).
- > Calling some platform intrinsic
    
    I'm also ok with that, with a little more explanation and best case with an example. Is there such a thing for all platforms?
- `spin_loop_hint` and similar functions, with the hint, that they may not work on all platforms.

**Comment by jyn514** [comment]
> spin_loop_hint and similar functions, with the hint, that they may not work on all platforms.

These won't work on *any* platform, they don't involve forward progress. @josephlr suggested changing them so that they emit some sort of intrinsic to LLVM, but there's no guarentee that change is going to be accepted; personally, it seems like solving this in the wrong place to me.

**Comment by jyn514** [comment]
>  Calling some platform intrinsic

I'm also dubious this will work, precisely because the bug is in LLVM and not the platform.

I think suggesting `-Z insert-sideeffect` on nightly and `read_volatile` elsewhere is unfortunately the best clippy can do currently.

**Comment by josephlr** [comment]
@jyn514 @flip1995 I opened https://github.com/rust-lang/rust/pull/77924 to make `spin_loop_hint` do what we want on all platforms (which will make it much easier to give good recommendations to users).

> I'm also dubious this will work, precisely because the bug is in LLVM and not the platform.

Here I was mainly thinking of my binary crates, where I call [`x86_64::instructions::hlt`](https://docs.rs/x86_64/0.12.2/x86_64/instructions/fn.hlt.html) in the panic handler. Obviously, that doesn't work for everyone. `cortex_m` users might call [`delay`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.delay.html), [`wfi`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.wfi.html), or [`nop`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.nop.html).

> I think suggesting `-Z insert-sideeffect` on nightly and `read_volatile` elsewhere is unfortunately the best clippy can do currently.

Ya I think the docs will basically say:
  - Don't `loop { }` because it can be UB and wastes CPU
  - If you can, `panic!` instead. This works everywhere and is better than just freezing up. Libraries should almost always do this.
  - Alternatively, on `std` targets:
    - Call `thread::sleep` or `thread::yield_now`
  - Alternatively, on `no_std` targets:
    - Call a target-specific intrinsic to sleep the thread (I'll give `x86_64::hlt` and a `cortex_m::wfi` as examples).
    - Call `sync::spin_loop_hint` (we would only recommend this if my change to `libcore` is accepted)
  - Finally, if you just want to work around the LLVM bug (but not the CPU issue):
    - For nightly, compile your code with `-Z insert-sideeffect`.
    - For stable, do a `read_volatile` on some local variable.

**Comment by Lokathor** [comment]
I think "an allocation" is over-stating it. your function's stack will be a single value bigger than normal, pretty trivial. It also inserts a single instruction into the loop instead of 0 instructions in the loop, which is again a fairly trivial difference, since the speed of the empty loop sure doesn't matter.

**Comment by josephlr** [comment]
@jyn514 @flip1995 I updated this PR to:
  - Make the lint a "correctness" lint
  - Improve the error messages
  - Improve the documentation as I laid out in https://github.com/rust-lang/rust-clippy/pull/6162#issuecomment-708062490

PTAL. Note that this will now make `empty_loop` a deny-by-default lint.

File: clippy_lints/src/loops.rs
Comment (flip1995) [inline clippy_lints/src/loops.rs:297]:
```suggestion
    /// these expressions can produce undefined behavior (in release mode). Also, these busy loops
```

File: clippy_lints/src/loops.rs
Comment (flip1995) [inline clippy_lints/src/loops.rs:319]: Should we also suggest the `llvm_asm!` version here for nightly? That way the user won't have to set any compiler flags and just has to change some code. Also the `llvm_asm!` version is a bit better than the `read_volatile` version, since it doesn't insert any code. (Even though the inserted code doesn't really make a difference performance wise or in any other way, since, as you stated, the program will be trapped in an infinite loop)

**Review by flip1995 — APPROVED** [review]
I'm overall satisfied with the changes, but would still appreciate a second pair of eyes, possibly with more `no_std` experience than me.

File: clippy_lints/src/loops.rs
Comment (jyn514) [reply clippy_lints/src/loops.rs:297]: I would not say 'can produce undefined behavior', I'd say 'can be miscompiled'. The user hasn't done anything wrong, the compiler is interpreting their code incorrectly.

File: tests/ui/empty_loop_no_std.rs
Comment (ebroto) [inline tests/ui/empty_loop_no_std.rs:15]: Should this be updated to match one of the current suggestions (`read_volatile`/`llvm_asm!`)?

**Review by ebroto — APPROVED** [review]
LGTM modulo flip1995/jyn514 comments

**Comment by therealprof** [comment]
I concur with the PR after the requested changes.

File: clippy_lints/src/loops.rs
Comment (jyn514) [inline clippy_lints/src/loops.rs:536]:
```suggestion
                let msg = "empty `loop {}` is currently miscompiled and wastes CPU cycles";
```

File: clippy_lints/src/loops.rs
Comment (flip1995) [reply clippy_lints/src/loops.rs:297]: I'm good with both or a combination "can be miscompiled, producing undefined behavior"

File: tests/ui/empty_loop_no_std.rs
Comment (josephlr) [reply tests/ui/empty_loop_no_std.rs:15]: I'll change it to `read_volatile`.

File: clippy_lints/src/loops.rs
Comment (flip1995) [reply clippy_lints/src/loops.rs:319]: Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

`-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

Comment (josephlr) [reply clippy_lints/src/loops.rs:319]: > Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

The old `asm!` got renamed to `llvm_asm!`. A new `asm!` was added (which is way better) that has a path to stablization.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

It becoming the default behavior is actually my concern (as then the flag would be removed).

Comment (flip1995) [reply clippy_lints/src/loops.rs:319]: https://github.com/rust-lang/rfcs/pull/2843 / https://github.com/rust-lang/rust/issues/70173

Ah `asm!` should be reimplemented, so it can become stable sometime in the future. Alright, in that case we should suggest to use `asm!`

Comment (therealprof) [reply clippy_lints/src/loops.rs:319]: > Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

Indeed. `llvm_asm!` is the new thing.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag or become enabled by default (which I really do not hope because it's just a bad and costly workaround), so whatever the stable long term solution will be, it's certainly not a `-Z` flag.

Comment (flip1995) [reply clippy_lints/src/loops.rs:319]: > The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag

Oh yeah, `-Z` flags are of course perma unstable, oops. Time to sleep 🛌

**Review by Lokathor — APPROVED** [review]

File: clippy_lints/src/loops.rs
Comment (josephlr) [reply clippy_lints/src/loops.rs:319]: I changed the recommendation to use an empty assembly statement `asm!("")`. It isn't marked `pure`, so this therefore a sideeffect.

**Comment by josephlr** [comment]
@flip1995 @ebroto @jyn514 All comments addressed. I think this is ready for merging.

I hope there won't be too much disruption from this lint going to deny-by-default.

File: clippy_lints/src/loops.rs
Comment (flip1995) [inline clippy_lints/src/loops.rs:319]: I would still include the hint to `-Zinsert-sideeffect`

**Review by flip1995 — APPROVED** [review]
Can you please squash the commits?

File: clippy_lints/src/loops.rs
Comment (josephlr) [reply clippy_lints/src/loops.rs:319]: Added

**Comment by josephlr** [comment]
> Can you please squash the commits?

Done, but the build is still broken until #6178 is merged.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Comment by jyn514** [comment]
So, after all that, https://github.com/rust-lang/rust/pull/77972 just landed. Maybe we no longer need to sound the alarm on this?

**Comment by ebroto** [comment]
IMO we could keep the more detailed documentation regarding the case of burning CPU, but I think the lint should go back to warn-by-default and ignore `no_std`

**Comment by jyn514** [comment]
In that case someone should r- before bors merges 😆

**Comment by ebroto** [comment]
@bors r-

(Not sure what triggered that last build 🤔)

**Comment by josephlr** [comment]
> IMO we could keep the more detailed documentation regarding the case of burning CPU, but I think the lint should go back to warn-by-default and ignore `no_std`

@ebroto @jyn514 I've updated this PR to do just this. But I've added a TODO to address #6161. I think we would ideally have this still trigger for `no_std` crates (as it's still bad style, in general). If we could somehow avoid firing this lint for `#[panic_handler]`, we could enable it everywhere else.

But this sort of discussion can be continued in #6161.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**
